### DISCUSSION TITLE
DatePicker Next 📅

### DISCUSSION THREAD

**Discussion by dakahn** [carbon-design-system/carbon#13355] [Ideas]
The NEXT Carbon DatePicker will be a from-the-ground-up refactor of the existing component that _should_ take into account the following:

## Dev
The primary motivation for the refactor is dropping the Flatpickr dependency improving DX, extensibility, and lowering maintenance cost for the component. Which leaves us with a few options in how to move forward (ordered naively in terms of difficulty): 
  - utilize a modern React date picker with relatively analogous APIs and functionality:
    - [DatePicker - React Spectrum](https://react-spectrum.adobe.com/react-spectrum/DatePicker.html)
    - [Airbnb React-Dates](https://github.com/react-dates/react-dates)
    - [Deque Cauldron Datepicker](https://dequeuniversity.com/library/aria/date-picker)
    - [Duet Design System DatePicker](https://github.com/duetds/date-picker)
  - utilize a DatePicker primitive as a base to build on top of:
    - [useDatePicker - React Aria](https://react-spectrum.adobe.com/react-aria/useDatePicker.html)
    - [Date Picker Dialog Example WAI-ARIA](https://www.w3.org/WAI/ARIA/apg/example-index/dialog-modal/datepicker-dialog.html)

    _This is arguably simpler than the previous option because there's less potential for conflict with existing CSS class names/styling_
  - Use browser native `<input type="date">`
      _All modern browsers have a built in implementation of a table calendar the user can select dates from. How different/alike they are in functionality and build is an unknown_
  - Build a wholly bespoke custom implementation 
  - ???

## Design
- the previous DatePicker design/UX was limited by the aforementioned Flatpickr dependency. With that dependency gone, we can evaluate if we want to revisit those designs or improve the existing UX
- product teams that are extending functionality can potentially have that functionality included out of the box with v12 DatePicker lifting up maintenance/support cost to the Carbon team
- ???

## Acessibility
It goes without saying that the v12 Carbon DatePicker should be designed/built with 100% WCAG AA level accessibility out of the box. We should lean on industry leaders in the accessibility space internal and external in that regard. Here are some relevant links:
- [Date and Time Pickers for All - Adobe](https://react-spectrum.adobe.com/blog/date-and-time-pickers-for-all.html)
- [Collecting Dates in an accessible way - Hassell Inclusion](https://www.hassellinclusion.com/blog/collecting-dates-accessible/)
- [Gov.uk design system Dates component discussion](https://github.com/alphagov/govuk-design-system-backlog/issues/43)
- [WC3 Datepickers docs](https://www.w3.org/WAI/PF/HTML/wiki/Datepickers)
- [A New Day: Making a Better Calendar - 24 Accessibility](https://www.24a11y.com/2018/a-new-day-making-a-better-calendar/)
- ???

## Compatibility
- The route we choose should be framework agnostic and able to be used within `@carbon/react` and `@carbon/web-components`

## Migration
A migration path from v11 to v12 should be considered which has the potential to include:
- an EXPERIMENTAL__DatePicker version we can test with Carbon users
- API mappings for previous props
- codemods for projects looking to update
- outreach to teams using and extending v11 DatePicker
- ???

**Comment by stuckless** [comment]
To minimize the case for needing to extend or create our own, please consider the following
1. if the component is going to display a date or allow the date to be entered from a keyboard it should expose parse/format function properties to allow teams to provide their own parsing/formatting of the date.
2. if the component is going to validate a date it would provide the ability to provide a custom validation function
3. Allow for custom placeholder text

**Reply by ondrakucera** [reply]
Yes, I'd like to see this as well. I can even imagine some of it could be improved even within the current version of DatePicker (but I understand priorities may lie elsewhere). For example, we are writing a new application where all localization of dates, times, numbers, etc. is done via Intl. The user selects their preferred language and we prepare Intl objects to format different data types. However, with DatePicker, unless I'm overlooking something, I cannot pass such instance to be used for formatting/parsing (nor can I pass a callback that would use it); there's just a `dateFormat` prop accepting a string format. In the end, we end up formatting dates differently within DatePicker and differently everywhere else.

**Comment by grahamharper** [comment]
Are there any more updates regarding plans for a new date picker?

The current one has so many issues, particularly accessibility issues. My team has applied a lot of customisations in a wrapper component to plug the holes but it's increasingly difficult and maintenance is very hard.

**Reply by sstrubberg** [reply]
hey @grahamharper thanks for reaching out. Unfortunately, no plans in the immediate future, but I appreciate you bringing it it up. The ol' datepicker is starting to show it's age and we'll need to address it soon.

**Reply by bdjos1** [reply]
Hey @sstrubberg, @tay1orjones,

I just wanted to follow up on this as we’re currently doing some forward planning on our UI roadmap and are trying to make informed decisions around the DatePicker usage and investment.

As mentioned above by Graham, we’ve had to apply local patches to the current DatePicker to address accessibility gaps (such as for #11800) and meet customer requirements. It would be helpful for us to better understand the Carbon team’s current thinking on it's future direction.

I completely understand that timelines are hard to commit to, but even a high-level signal on intent or priority would really help inform our planning.

Thanks

**Comment by tay1orjones** [comment]
* https://github.com/gpbl/react-day-picker
* https://github.com/Hacker0x01/react-datepicker

**Comment by boeckMt** [comment]
What is about the [carbon web-components](https://github.com/carbon-design-system/carbon/tree/main/packages/web-components), are there also plans to update?

**Reply by sstrubberg** [reply]
As it stands today and for the foreseeable future, any work we take on, and this includes future DatePicker work, we would be build for both React and Web Components.

**Comment by tay1orjones** [comment]
One additional requirement that would be nice to have as part of this is multiple calendar support, https://github.com/carbon-design-system/carbon/issues/7265. This was [recently brought up](https://ibm-studios.slack.com/archives/C2K6RFJ1G/p1758110085918049) again internally.

**Comment by tay1orjones** [comment]
Another nice to have would be combining datepicker and timepicker, https://github.com/carbon-design-system/carbon/issues/10186. Probably not to actually merge the components (we could, I just don't think it's necessary?), but we could integrate them together more intentionally through composability.

**Comment by bdjos1** [comment]
Hi @tay1orjones,

Just wondering if there has been any decision made yet on the future direction with respect to the Date Picker?

I completely understand that timelines are hard to commit to, but even a high-level signal on intent or priority would really help inform our planning.

Thanks

**Comment by sangeethababu9223** [comment]
We recently faced an issue where the date-range picker web component was throwing `Uncaught TypeError: fp.getDateStr is not a function`. To fix this, we had to [patch flatpickr to expose internal methods and address Shadow DOM compatibility issues](https://github.com/carbon-design-system/carbon/pull/18022), then [bundle the entire library](https://github.com/carbon-design-system/carbon/pull/22081) with web components to ensure these patches work in production.
